src/network/connection.cpp

```diff
@@ -62,18 +62,31 @@ namespace con
 
 #define PING_TIMEOUT 5.0
 
-BufferedPacket makePacket(Address &address, const SharedBuffer<u8> &data,
+u16 BufferedPacket::getSeqnum() const
+{
+	if (size() < BASE_HEADER_SIZE + 3) {
+		//std::cout << (int)data[0xFFFFFFFFF] << std::endl;
+		throw ConnectionException("Packet too short to contain a seqnum");
+	}
+
+	return readU16(&data[BASE_HEADER_SIZE + 1]);
+}
+
+BufferedPacketPtr makePacket(Address &address, const SharedBuffer<u8> &data,
 		u32 protocol_id, session_t sender_peer_id, u8 channel)
 {
 	u32 packet_size = data.getSize() + BASE_HEADER_SIZE;
-	BufferedPacket p(packet_size);
-	p.address = address;
+	auto p = std::make_shared<BufferedPacket>(packet_size);
+	p->address = address;
+
+	u8 *packet = p->ptr_write();
 
-	writeU32(&p.data[0], protocol_id);
-	writeU16(&p.data[4], sender_peer_id);
-	writeU8(&p.data[6], channel);
+	writeU32(&packet[0], protocol_id);
+	writeU16(&packet[4], sender_peer_id);
+	writeU8(&packet[6], channel);
 
-	memcpy(&p.data[BASE_HEADER_SIZE], *data, data.getSize());
+	memcpy(&packet[BASE_HEADER_SIZE], *data, data.getSize());
+	p->lock();
```

Comment:
I wonder if this could be made to work nicer using the "const shared ptr" I proposed, looks like `lock()` is always called directly after creation.